### PR TITLE
email subscription

### LINKED ISSUE #4493: email subscription (newsletter, features, security updates)
We'd like to give lakeFS users the possibility to subscribe to newsletters, features, and security updates.

In order to do so, we'd like in the setup phase to add a preliminary step where users will be asked to provide an email address (no need to verify its legitimacy)

Email can be saved in the local database for future use (i.e unsubscribe) and sent (anonymized) for stats.

Mock:
image

### PR DESCRIPTION
**Pull request by eladlachmi**
Resolves #4493

### CODE REVIEW

File: webui/src/pages/setup/communicationPreferences.tsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+                            <Form.Group controlId="updates-check" className="mt-4">
+                                <Form.Label>News & feature updates</Form.Label>
+                                <Form.Check type="checkbox" checked={updatesCheck} onChange={handleUpdatesChange} label="I'd like to receive news and feature updates" />
+                            </Form.Group>
+
+                            <Form.Group controlId="security-check" className="mt-4">
+                                <Form.Label>Security updates</Form.Label>
+                                <Form.Check type="checkbox" checked={securityCheck} onChange={handleSecurityChange} label="I'd like to receive security updates" />
+                            </Form.Group>
```
Comment (ortz) [inline webui/src/pages/setup/communicationPreferences.tsx:66]: I think it's better to have the security updates subscription checked by default

File: webui/src/pages/setup/communicationPreferences.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                        <Card.Text>
+                            Providing your email address is optional.
+                            We recommend at least signing up for security updates. <br />
+                            You may unsubscribe from either or both at any time.
+                        </Card.Text>
+                        <Form onSubmit={submitHandler}>
+                            <Form.Group controlId="user-email" className="mt-4">
+                                <Form.Label>Email (optional)</Form.Label>
+                                <Form.Control type="email" placeholder="name@company.com" value={userEmail}  onChange={handleEmailChange} autoFocus/>
+                            </Form.Group>
```
Comment (ortz) [inline webui/src/pages/setup/communicationPreferences.tsx:56]: I wonder if we'd like to allow empty value here...
let me check that.

**Review by ortz — CHANGES_REQUESTED** [review]
Overall, looks really good!
Some recommendations and questions...

File: pkg/auth/metadata.go [hunk 3]
```diff
@@ -12,8 +14,20 @@ import (
 )
 
 const (
-	InstallationIDKeyName = "installation_id"
-	SetupTimestampKeyName = "setup_timestamp"
+	InstallationIDKeyName  = "installation_id"
+	SetupTimestampKeyName  = "setup_timestamp"
+	CommPrefsSetKeyName    = "comm_prefs_set"
+	EmailKeyName           = "hashed_user_email"
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/auth/metadata.go:20]: As discussed above, it is not hashed.

File: webui/src/pages/setup/communicationPreferences.tsx (hunk 1 above)
Comment (eladlachmi) [reply webui/src/pages/setup/communicationPreferences.tsx:66]: At first, I had it checked by default, but it's more aligned with current regulations to leave it unchecked. Checking the boxes is a more clear-cut opt-in action.  
If they want to receive these updates, they'll check the box.

File: pkg/api/controller.go [hunk 4]
```diff
@@ -3302,6 +3306,65 @@ func (c *Controller) Setup(w http.ResponseWriter, r *http.Request, body SetupJSO
...
+		writeError(w, http.StatusPreconditionFailed, "wrong step - CommPrefs have already been set")
+		return
+	}
+
+	response := NextStep{
+		NextStep: string(auth.SetupStateCommPrefsDone),
+	}
+	// user chose not to provide their email address,
+	// update that this step is skipped, and move on to the next step
+	if len(*body.Email) == 0 {
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/api/controller.go]: That'll work. Fixed.

File: pkg/stats/sender.go [hunk 5]
```diff
@@ -96,6 +97,34 @@ func (s *HTTPSender) SendEvent(ctx context.Context, installationID, processID st
 	return nil
 }
 
+func (s *HTTPSender) UpdateCommPrefs(ctx context.Context, email, installationID string, featureUpdates, securityUpdates bool) error {
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/stats/sender.go]: I did not until you mentioned it. Fixed.

File: pkg/stats/collector.go [hunk 6]
```diff
@@ -338,6 +346,15 @@ func (s *BufferedCollector) CollectMetadata(accountMetadata *Metadata) {
 	}
 }
 
+func (s *BufferedCollector) CollectCommPrefs(email, installationID string, featureUpdates, securityUpdates bool) {
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/stats/collector.go]: Fixed.

File: pkg/auth/metadata.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/auth/metadata.go]: Fixed.

File: webui/src/pages/setup/index.jsx [hunk 7]
```diff
@@ -41,6 +41,10 @@ const SetupContents = () => {
     }, [setDisabled, setSetupError, setSetupData, setup]);
 
     const onSubmitCommunicationPreferences = useCallback(async (userEmail, updatesChecked, securityChecked) => {
+        if (!userEmail) {
+            setSetupError("Email is require.");
```
Comment (ortz) [inline webui/src/pages/setup/index.jsx:45]: required*

**Review by ortz — APPROVED** [review]
One small comment, other than that, approved!

File: webui/src/pages/setup/index.jsx (hunk 7 above)
Comment (eladlachmi) [reply webui/src/pages/setup/index.jsx]: Fixed

File: pkg/auth/metadata.go [hunk 8]
```diff
@@ -24,6 +38,9 @@ func init() {
 
 type MetadataManager interface {
 	IsInitialized(ctx context.Context) (bool, error)
+	GetSetupState(ctx context.Context, emailSubscriptionEnabled bool) (SetupStateName, error)
+	UpdateCommPrefs(ctx context.Context, commPrefs CommPrefs) (string, error)
+	UpdateSkipCommPrefs(ctx context.Context) error
```
Comment (guy-har) [inline pkg/auth/metadata.go:43]: `UpdateSkipCommPrefs` isn't used

**Review by guy-har — APPROVED** [review]
Looks great 👍 
Minor comments

File: pkg/api/controller.go [hunk 9]
```diff
@@ -3298,6 +3302,63 @@ func (c *Controller) Setup(w http.ResponseWriter, r *http.Request, body SetupJSO
...
 }
 
+func (c *Controller) SetupCommPrefs(w http.ResponseWriter, r *http.Request, body SetupCommPrefsJSONRequestBody) {
+	ctx := r.Context()
+	emailSubscriptionEnabled := c.Config.IsEmailSubscriptionEnabled()
+	initialized, err := c.MetadataManager.GetSetupState(ctx, emailSubscriptionEnabled)
+	if err != nil {
+		writeError(w, http.StatusInternalServerError, err)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/api/controller.go]: Fixed

File: pkg/auth/metadata.go [hunk 10]
```diff
@@ -91,6 +182,30 @@ func (m *KVMetadataManager) UpdateSetupTimestamp(ctx context.Context, ts time.Ti
...
 }
 
+func (m *KVMetadataManager) UpdateCommPrefs(ctx context.Context, commPrefs CommPrefs) (string, error) {
+	encodedEmail := ""
+	var err error
+	if commPrefs.UserEmail != "" {
+		encodedEmail = b64.StdEncoding.EncodeToString([]byte(commPrefs.UserEmail))
+		if err != nil {
+			return "", err
+		}
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply pkg/auth/metadata.go]: Remnants from when it was hashed, which did return an error. Fixed.

File: pkg/auth/metadata.go (hunk 8 above)
Comment (eladlachmi) [reply pkg/auth/metadata.go]: Removed.

**Pull request merged**